src/lib.rs

```diff
@@ -75,179 +77,105 @@ pub struct Neighbor<Unit, Ix = usize> {
     pub distance: Unit,
 }
 
-/// Implement this trait on data structures (or wrappers) which perform KNN searches.
-pub trait Knn {
-    type Ix: Copy;
-    type Point;
-    type Metric: Metric<Self::Point>;
-    type KnnIter: IntoIterator<
-        Item = Neighbor<<Self::Metric as Metric<Self::Point>>::Unit, Self::Ix>,
-    >;
-
-    /// Get `num` nearest neighbors of `target`.
-    ///
-    /// For many KNN search algorithms, the returned neighbors are approximate, and may not
-    /// be the actual nearest neighbors.
-    fn knn(&self, query: &Self::Point, num: usize) -> Self::KnnIter;
-
-    /// Get the nearest neighbor of `target`.
-    ///
-    /// For many KNN search algorithms, the returned neighbors are approximate, and may not
-    /// be the actual nearest neighbors.
-    #[allow(clippy::type_complexity)]
-    fn nn(
-        &self,
-        query: &Self::Point,
-    ) -> Option<Neighbor<<Self::Metric as Metric<Self::Point>>::Unit, Self::Ix>> {
-        self.knn(query, 1).into_iter().next()
-    }
+pub trait KnnFromBatch<B> {
+    fn from_batch(batch: B) -> Self;
 }
 
-/// This trait gives knn search collections the ability to give the nearest neighbor points back.
+/// This trait gives knn search collections the ability to give the nearest neighbor values back.
 ///
 /// This is not the final API. Eventually, the iterator type will be chosen by the collection,
 /// but for now it is a [`Vec`] until Rust stabilizes GATs.
-#[cfg(feature = "alloc")]
-pub trait KnnPoints: Knn {
+pub trait Knn<'a> {
+    type Ix: Copy;
+    type Point: 'a;
+    type Value: 'a;
+    type Metric: Metric<Self::Point>;
+    type KnnIter: Iterator<
```

Comment:
This should be `IntoIterator` rather than `Iterator`. The reason is that then the implementation can return, for instance, a `Vec`, and the caller can use the `Vec` rather than just the iterator. This is done in some of my downstream libraries.

Reply:
Also, iterators also implement `IntoIterator`, so the implementation can still choose to return that if it wants to.